src/components/App.js

```diff
@@ -44,618 +35,119 @@ class App extends Component {
-  }
-
-  setWsConnection() {
-    let websocketUrl = this.getWebsocketUrl();
-    let wsConnection = this.newWebSocket(websocketUrl);
-
-    // Kill previous wsConnection connection
-    if (this.state.wsConnection) {
-      this.state.wsConnection.close();
-    }
-
-    this.setState({
-      wsConnection: wsConnection
+      showUsernameModal: true
     });
   }
 
-  keypairFromURLHash(urlHash="") {
-    urlHash = urlHash || document.location.hash;
-
-    if (this.pincodeRequired(urlHash)) {
-      this.promptForPincode();
-      return;
-    }
-
-    let { passphrase, isNewPassphrase } = getPassphrase(urlHash);
-
-    if (isNewPassphrase) {
-      document.location.hash = '#' + passphrase;
-      this.displayAlert('New room created!', 'success');
-    }
-
-    console.log("URL hash is `%s`", passphrase);
-
-    let email = getEmail(passphrase);
-
-    miniLock.crypto.getKeyPair(passphrase, email, (keyPair) => {
-      // Code from https://github.com/kaepora/miniLock/blob/ffea0ecb7a619d921129b8b4aed2081050ec48c1/src/js/ui.js#L78
-      // May be useful:
-      // https://github.com/kaepora/miniLock/blob/master/src/js/miniLock.js#L18
-      miniLock.session.keys = keyPair
-      miniLock.session.keyPairReady = true
-
-      let mID = miniLock.crypto.getMiniLockID(keyPair.publicKey);
-      console.log("mID ==", mID);
-
-      this.setState({
-        keyPair: keyPair,
-        keyPairReady: true,
-        mID: mID
-      }, this.login);
-    })
-  }
-
-  paranoidMode(urlHash=document.location.hash) {
-    return urlHash.endsWith('----');
-  }
 
-  pincodeRequired(urlHash=document.location.hash) {
-    return urlHash.endsWith('--');
-  }
-
-  promptForPincode() {
-    this.setState({
-      showPincodeModal: true
-    })
-  }
-
-  onClosePincodeModal() {
+  onClosePincodeModal = () => {
     this.setState({
       showPincodeModal: false
     });
   }
 
-  onSetPincode(pincode="") {
+  onSetPincode = (pincode = "") => {
     if (!pincode || pincode.endsWith("--")) {
       this.onError('Invalid pincode!');
       return;
     }
-
-    var urlHash = document.location.hash + pincode;
-    this.keypairFromURLHash(urlHash);
-
-    this.onClosePincodeModal();
-
-    if (this.paranoidMode()) {
-      this.props.setUsername(PARANOID_USERNAME);
-    } else {
-      this.promptForUsername();
-    }
+    this.props.initConnection(pincode);
   }
 
-  onCloseUsernameModal() {
+  onCloseUsernameModal = () => {
     this.setState({
       showUsernameModal: false
     });
   }
 
-  toggleInfoModal = () => {
-    this.setState((prevState) => {
-      return { showInfoModal: !prevState.showInfoModal }
-    })
-  }
-
-  onSetUsername(username) {
-    if (!username) {
-      this.onError('Invalid username!');
-      return;
-    }
-
-    if (!this.paranoidMode()) {
-      localStorage.setItem(USERNAME_KEY, username);
-    }
-
-    this.props.setUsername(username);
-
+  handleSetUsername = (username) => {
     this.setState({
```

Comment:
@Spetastian `previousUsername`'s value is indeed now being grabbed from localStorage, but the user's username isn't being stored there in the first place :-).

Would this line be a good place to do `localStorage.setItem(USERNAME_KEY, username)`, or I suppose that should be in `setUsername`...

Reply:
Since its a side effect that talks to the host OS, that is a part of the Rx flow in `chatEpics`